### DISCUSSION TITLE
Request documentation!

### DISCUSSION THREAD

**Discussion by danharrin** [filamentphp/filament#876] [General]
Hey all!

Filament is relatively new, so documenting every feature is an ongoing process.

I'd like to ask you - the community - to tell us exactly what you're missing from the docs.

I'll aim to address all replies with a link to the docs, a video tutorial, or a solution reply if it's too niche :)

Go! 👇

**Comment by nicolaibaaring** [comment]
I have found these topics that don't seem to be documented.

- Custom modals on pages in the admin panel
- Best practice for implementing relations' tables on view pages in the admin panel
- Custom filters, which were present in v1, are not in the documentation for table builder
- ButtonAction's openUrlInNewTab is not found in docs. It's, though, pretty self-explanatory when viewed in CanOpenUrl trait.

I'm still curious for the best practice on the first two topics.

**Reply by ryangjchandler** [reply]
@nicolaibaaring You should be able to find information about custom filters here: https://filamentadmin.com/docs/2.x/tables/filters

**Reply by nicolaibaaring** [reply]
@ryangjchandler Then I should have been more specific here. 

It's about named reusable custom filters that I had a brief discussion about with @danharrin on the Discord. In v1 it was documented that you could make your own filter classes like `DescFilter`, but in v2 all filters take a name in the static constructor. This means that even named filters have to take a name as argument even though the name is already applied in the filter class. `DescFilter::make('desc')` doesn't seem like the right way to go, but the constructor is final, so it's kinda forced. 

There is no documentation about these in v2 docs. 

I have made a couple of initial suggestions on how it could be changed - one being to add another static constructor and do `DescFilter::new()`.

**Reply by ryangjchandler** [reply]
Right, I see. The name isn't technically applied in the filter class though unless you're overwriting a method (`__construct` is `final` so it can't be overwritten) -- the name is used to store the filter values on the Livewire component. I don't really see a problem with having to define that because your `Filter` class might be something generic like `TextFilter` where the name really could be ambiguous.

**Reply by nicolaibaaring** [reply]
OK. Then I see too. That's definitely another perspective than I had in mind. I had that reusable filters had generic names and thus the applied name would be defining the filter in general. Then I guess nothing needs to be changed and we can just use the built in names of custom classes for cases where they don't need more specific names. Thanks.

**Reply by danharrin** [reply]
Yeah, when I replied on Discord earlier this also slipped my mind. Custom filter state is recorded in a property with the name of the filter, so names are required in v2.

**Reply by ryangjchandler** [reply]
The alternative approach here would be defaulting the `$name` property to the current `static::class`.

**Reply by MACscr** [reply]
I second the request for more documentation for modals. I really want them for both page action buttons and table row actions. I know its possible, just not as straight forward as it should be in my opinion. To me, if its simply a modal for editing a record, shouldn't that same modal be able to be used pretty much anywhere since just a record needs passed to it? Same really with creation. Its simply a form builder instance in a modal.

**Comment by ghost** [comment]
I'd like to add that hiding/showing of LinkAction or any action on table is not yet added on the docs (or did I miss something?). But that's already answered here on the discussion. :)

**Reply by danharrin** [reply]
Thanks! Add this to my docs todos.

**Comment by titantwentyone** [comment]
Tests! Please :) What to test and how would be great. I think (although I can't find the source) that there was mention of better testing available in v2.

**Reply by danharrin** [reply]
Yup - this is on my todo list to document. If you'd like to see some example tests for admin panel pages right now, check out https://github.com/laravel-filament/filament/tree/2.x/tests/src/Admin/Resources/Pages

**Reply by rompetomp** [reply]
Hi @danharrin, is it possible to also include example tests for filament forms in the documentation?

**Comment by pxlrbt** [comment]
Best practices for plugin development. What's the best way to let users customize as much as possible.

**Reply by danharrin** [reply]
Added to my todos 👍

**Comment by tangjh** [comment]
1) how about a complete document/tutorial on how to create a new field eg color picker, how to register js, css, etc.
same topic but development from both "plugin" and "add on" perspective.
can compare pros and cons.

2) another topic can be on performance tuning on queries when loading data for table columns, forms multi-select, checkboxes with relations.  eg how to do eager loading.

**Reply by tobz-nz** [reply]
Big thumbs up for 1. building custom fields - the docs currently are next to nothing, and the process not very intuitive at all - making custom fields very difficult at the moment.

**Reply by zepfietje** [reply]
You may want to check this tutorial series by @ryangjchandler: https://ryangjchandler.co.uk/posts/building-a-custom-color-palette-field-in-filament-part-0.

**Reply by tobz-nz** [reply]
@zepfietje Thanks - having a look now.
The main issue I'm having is just setting the state - like how the hell do you do it?

I've basically copied the base text component, then trying to add some custom logic on `@input` which looks up some data from an API which I then need to save along with the original input value. For the life of me I can't figure out how to store the extra data or even just manipulate data before saving it (in the field, not in the form or model as this field will be used across multiple models)

**Reply by zepfietje** [reply]
That's explained in the docs: https://filamentphp.com/docs/2.x/forms/fields#building-custom-fields.

Also, the generated custom field view contains this comment: "Interact with the `state` property in Alpine.js".

**Reply by tobz-nz** [reply]
@zepfietje yeah... I just get errors like `Unable to set component data. Public property [$state] not found on component: [app.filament.resources.person-resource.pages.edit-person]`

simplified template snippet:

```php

<div class="flex-1" x-data="{ state: $wire.entangle('{{ $getStatePath() }}') || {} }">
            <input
                {{ $applyStateBindingModifiers('wire:model') }}="state.url"
                type="{{ $getType() }}"
                id="{{ $getId() }}"
                @input.debounce="state.data = {someValue: '....'}"
>
</div>

**Reply by tobz-nz** [reply]
@zepfietje I finally got it working - thanks for pointing me in the right direction. This whole thing does need a LOT better documentation. I had it in my head that Fields were Livewire components - which they are not.

**Reply by pxlrbt** [reply]
> @zepfietje I finally got it working - thanks for pointing me in the right direction. This whole thing does need a LOT better documentation. I had it in my head that Fields were Livewire components - which they are not.

Since you just went through the process and know the pain point, do you wanna PR some additions to those docs?

**Reply by tobz-nz** [reply]
@pxlrbt I would, but I really don't think I have a good enough understanding of how to do it properly to be honest.

**Reply by zepfietje** [reply]
Even if you'd open a PR with some basic notes of info missing according to you, that'd be of help. Then we can take it from there.

**Comment by titantwentyone** [comment]
Is there anywhere we can help add to the documentation? Just found out about the `extraInputAttributes` method for forms and thought the docs could have this added as well as clarification on the difference between this and `extraAttributes`. I'm actually doing a lot of "noseying" around the source and finding lots of useful nuggets! I saw there is repo for `laravel-filament/filamentadmin.com` but think this is still a WIP?

**Reply by ryangjchandler** [reply]
@titantwentyone Each of the respective packages in this mono-repo has a `docs` folder that you can edit & PR too.

**Reply by titantwentyone** [reply]
Fantastic. Thank you!

**Comment by ManojKiranA** [comment]
Detailed Docs for View Page. Curently  its just tells how to [create a page](https://filamentadmin.com/docs/2.x/admin/resources#view-page).

Even in demo repo there is no view page either

**Reply by nicolaibaaring** [reply]
I came here to write the same. It's like the documentation just has a full gap where this could belong. 
Given the source code, it looks like the ViewRecord is just supposed to display a disabled "form", which I, in that case, would assume would be display values and labels (not form fields or disabled form fields). 
It would also be a reasonable approach to build the display as a form with preconstructed display "fields" - using the form schema.

But I actually do not see how this is intended besides using the actual record model and constructing your own view, which then does not benefit from layouts, tables and more.

How is it intended and will you add it to the documentation?

**Comment by ArtMin96** [comment]
There is no sample for URL dynamic parameters on the [Pages](https://filamentphp.com/docs/2.x/admin/pages) page. I did the following and it worked. I do not know if there is a better option or not.

Page class:
```php
protected static ?string $slug = '/custom-url/{record}';
```

Resource class:
```php
IconButtonAction::make('view')
                    ->label('View record')
                    ->url(fn (Model $record): string => url(Page::getUrl(['record' => $record])))
                    ->icon('heroicon-o-eye')
                    ->color('dark'),
```

**Comment by frikishaan** [comment]
It will be better to have some screenshots, especially in the sections like [Layouts](https://filamentphp.com/docs/2.x/forms/layout).

**Reply by pxlrbt** [reply]
I agree that screenshots would be nice. But if we start adding screenshots, people will probably want them for other parts, too. That adds a lot of maintenance burden for the docs.

**Reply by kurucu** [reply]
What about, instead of screenshots, sort of blueprints? Like dotted frames that show you what will happen.

**Comment by suhaimim** [comment]
Further guide (better with sample) for below topics;

1. Custom page view with model value
2. Custom cascade dropdown select 
3. Custom auto complete field (Docs exist! Pls. ignore)
4. Export/Print to PDF
5. Edit page view/edit permission for some part of fields 

would be nice for beginner like me.

**Comment by pxlrbt** [comment]
1. `->visible()/->hidden()` was removed I think (https://filamentphp.com/docs/2.x/admin/resources/getting-started#hiding-components-based-on-the-page)
2. `->suffixAction()/->prefixAction()` needs documentation
3. `->createOptionForm()` is not documented.

**Reply by danharrin** [reply]
`createOptionForm()` is documented

**Reply by pxlrbt** [reply]
Yeah, found it. Code blocks are apparently not searchable, so when searching for `createOptionForm()` there are no results 😅

**Reply by oliwerhelsen** [reply]
@pxlrbt where did you find it? :-). I'm can't seems to get any of the Closure variables to work in createOptionUsing. I hope for this to be documented under the same page where you found the createOptionForm.

**Reply by pxlrbt** [reply]
@oliwerhelsen https://filamentphp.com/docs/2.x/forms/fields#creating-new-records
For support it's best to use Discord or the GH Discussions.

**Comment by pxlrbt** [comment]
I cannot find`Group` and especially `Group::make()->relationship()` in the docs.

**Reply by danharrin** [reply]
Group isn't there, but https://filamentphp.com/docs/2.x/forms/layout#saving-data-to-relationships is

**Comment by aurawindsurfing** [comment]
@danharrin maybe it will make sense to let the community do pull requests to the documentation? Similar to what laravel docs are allowing https://github.com/laravel/docs

**Reply by danharrin** [reply]
You can already do that, each package has a /docs directory

**Reply by aurawindsurfing** [reply]
Nice one! I was looking in a top level directory for it ;-)

**Comment by Thiktak** [comment]
Hello,

I didn't find documentation about two topics:

## Table Header/Heading
Example: add title to Finalement Table card, with $this->table->heading('My title');

## Actions linked to the table/header.
protected function getTableHeaderActions(): array

![image](https://user-images.githubusercontent.com/1201486/183254962-dbbef2de-dd30-4605-b77f-2ef25466c8f7.png)

**Comment by eugenefvdm** [comment]
I've been watching Povilas's [training video on Filament](https://laraveldaily.teachable.com/p/laravel-filament-admin-practical-course) and in episode `Table Footer: Total Column Calculations` he uses the below method in List Resource which I couldn't find documented anywhere:

```php
protected function getTableContentFooter(): ?View
    {
        return view('...');
    }
```

It goes on to show exactly how to create a blade with all the right Tailwind classes..I thought this could be useful documentation.

**Reply by maccarib** [reply]
I've been trying hard to achieve something like that feature mentionned. Thanks for this hint! I nevertheless don't have access to that video. Could you give me a hint on how the access to the table data works in that case? I hope I can manage to work out all the other parts.

**Reply by Saifallak** [reply]
+1

**Reply by dream-hun** [reply]
It works in V2 but It doesn't work in latest version i mean V3

**Reply by danharrin** [reply]
https://filamentphp.com/docs/3.x/tables/summaries

**Comment by KaanTanis** [comment]
->extraModalActions

I didn't see it anywhere. If it doesn't already exist, let's add it. With other actions class details :)

![image](https://user-images.githubusercontent.com/64961969/183888534-26510018-9aad-4bf9-85e6-a15c63b5368b.png)

**Reply by tomasvanrijsse** [reply]
I tried to implement this extraModalActions and was able to make the button show up.
But I had no clue what to do with `action()` and therefore couldn't make a function that would run when the button was clicked.
@KaanTanis do you have a clue what to enter as the first paramter of this action method?

**Comment by SquidwardHill** [comment]
**Small Potato**: This is not quite a request for documentation, but related:

It would improve usability of the documentation a lot if you had either a "Back to Top" or "Next Topic" button at the bottom of each topic page (or both).  

Thanks for all that you all do!

**Comment by lutfifirdaus** [comment]
Pls add information about how registering resource's custom page that need to add on top of 'view' and 'edit' page, cuz i recently got this problem and me not found any of that on custom page section. 

I also wanna say thanks to this package team and participation that make this great package !!!

**Reply by nickspringham** [reply]
I know this is late to the party but I found how to do this quite difficult as well. I did manage to figure out that you can add a new action to the $table->actions array which will then take you to the custom page you created (my custom page is 'Review')

```
->actions([
    Tables\Actions\EditAction::make(),
    Tables\Actions\Action::make('Review')
        ->color('success')
        ->url(fn(AppReview $record): string =>  self::getUrl('review', ['record' => $record]))
])
```

Hope that helps anyone who might come accross this in the future 👍

**Comment by dissto** [comment]
Components should have a table underneath showing all available properties/options with a little description.

For example for the TextInput

- autofocus
- required
- string
- etc.

**Comment by dissto** [comment]
The algoilia search modal should be available in dark mode as well

**Comment by eduardoarandah** [comment]
@danharrin I think I've simplified the process for theming **How to contribute to documentation?**

Specifically this part:

- https://filamentphp.com/docs/2.x/admin/appearance#building-themes 

I searched the repo and it looks like it's data-driven 
https://github.com/filamentphp/filamentphp.com/blob/main/app/Http/Controllers/DocumentationController.php

**Reply by danharrin** [reply]
Each package in this repo has a /docs directory

**Comment by RX-77IT** [comment]
I don't know if it's intended, but the method `getStateUsing(fn ($record) => ... )` to change a Column state is cited only on the "[Upgrading from v1.x](https://filamentphp.com/docs/2.x/admin/upgrade-guide#columns)" page.

**Comment by OvalMedia** [comment]
It would be nice if the source code (functions) was documented. I would assume that a lot of questions here would answer themselves if the code was documented. Also it helps when you work with an IDE like phpstorm.

**Comment by RX-77IT** [comment]
Documentation about `mutateRelationshipDataBeforeCreateUsing()` and `mutateRelationshipDataBeforeSaveUsing()` maybe? 
I just found out them from [here](https://github.com/filamentphp/filament/discussions/4133).

**Comment by tobz-nz** [comment]
The main thing is - I think - an actual example of setting a value (multiple values, as that is often a requirement) using Alpine.
Currently is just says “use alpine” - not all that useful IMHO.

Also examples and how to do hydration, mutation, etc.. in the field class (not in the instantiation of the class) - if that is even possible?

..and is it possible to call actions on the field via alpine events (like livewire actions)? if it is examples of how.

I think those few things would go a long way to making it much easier to build custom fields.

**Comment by Rudston** [comment]
**Documentation concerning Select with multiple() enabled please:**
- I cannot see any way to get the current values to be displayed (usually achieved with fill(), default() also of no use)
- I also have the issue (bug?) that items can be selected twice

I am new to GitHub use and was about to write issues for these, but this requires setting up a whole repo just for the sake of one field:

`Select::make('tagIds')
                        ->multiple()
                        ->label('Tags')
                        ->options(VideoTag::all()->sortBy('name')->pluck('name', 'id'))
                        ->getOptionLabelsUsing(fn ($value): ?string => VideoTag::find($value)?->name)
                        ->searchable()`
Values can be selected and saved - just those two issues

**Reply by danharrin** [reply]
Please send what you've tried for filling the select. It should be the same as normal when using `$this->form->fill([])`.

Are you also experiencing #2363?

**Reply by Rudston** [reply]
Hello, thanks for speedy reply. I am using the usual fill as above:
`
$this->form->fill(
            [
                'date' => $this->video->webinar_date,
                'title' => $this->video->title,
                'description' => $this->video->description,
                'vimeoUrl' => $this->video->vimeo_url,
                'chatUrl' => $this->video->chat_url,
                'tagsIds' => $this->video->tags->pluck('id')->toArray()
            ]
        );
`
All the other fields are filled correctly, just the multi select not: though it does work for selection purposes. 
The field is coded as follows:

`public $tagIds =[];
		  ...

Select::make('tagIds')
                        ->multiple()
                        ->label('Tags')
                        ->options(VideoTag::all()->sortBy('name')->pluck('name', 'id'))
                        ->searchable(),`

#2363 - is not occurring - sorry, that was an error in my data. 

Note: I am using the simplest approach; not 'registering' a model e.g.; must be doing something basic incorrectly as I have the same problem using a checkbox list - saves correctly, but does not fill the existing values on loading the form.

Hope that helps! I'm really stuck - filament is otherwise workday really well every else its in use...

**Reply by danharrin** [reply]
This does work on my end, can you provide a reproduction repository for me please in a new issue?

**Reply by Rudston** [reply]
Hi, Sorry, that would take me a while - I would have to check with my client once he returns from vacation.
    In the meantime, I can indicate:
    - php version - 8.0
    - laravel version - 9.43
    - livewire version - 2.10
    - filament version - 2.16

**Reply by danharrin** [reply]
You don't need to share the client repo, you create a new one just reproducing this issue. Otherwise it'll be very bloated with everything else we don't need to know about.

**Reply by Rudston** [reply]
Hi Dan, I have created a reproduction repo - see "Multi-select does not display current values on fill #5670". Hope that helps!

**Reply by Rudston** [reply]
Hi Dan, This may be totally irrelevant, but when dealing with an issue with another custom filament field I discovered that the field set a listener on alpine:init, but alpine had already been initialised before the component was setup so that the listener as never called ... Only by ensuring that Alpine.start() was called at the very end, would the component concerned function. My layout and or my app.js could be incorrect.

**Reply by elmergustavo** [reply]
thank

**Comment by mstfkhazaal** [comment]
Please add each category that can be changed into themes:
like: 
.filament-main-topbar ,.filament-modal-content,.filament-modal-window,
.filament-sidebar-header ,.filament-sidebar-nav,.filament-body,.filament-app-layout,.filament-stats-card

I recommend that there be a YouTube channel for this tool

**Comment by dogeow** [comment]
Tutorial, like `React Tutorial`  or `Next.js Tutorial`.

This is just an introductory tutorial for basic development, just put a link for other advanced content.

This ensures that we will not get stuck halfway.

**Comment by KONEY** [comment]
Notifications docs are ok for the basic part but the broadcast section is confusing because it's no more clear if it's about the same kind of (flash?) notifications or what.

**Reply by danharrin** [reply]
Broadcast notifications are a Laravel concept, which is why we have not explained it again. Will see if we should expand this documentation more.

**Reply by KONEY** [reply]
Of course and there's an implementation of them in Filament, its docs could be improved.

**Reply by danharrin** [reply]
What specifically is missing from the docs?

**Reply by KONEY** [reply]
Not sure I can speak about something missing but following the flow of information doesn't add up:

1) https://filamentphp.com/docs/2.x/notifications/installation says how to send a notification providing a visual feedback: https://user-images.githubusercontent.com/44533235/180972526-da7ebb2f-24f6-4cf4-8e4c-d261574384ab.jpg

2) Step to https://filamentphp.com/docs/2.x/notifications/broadcast-notifications where one would expect to find out how to apply point 1 to multiple users but despite the same name these are not the same GUI notifications as before but browser push-notification.

Also in the same section this explanation is a bit confusing: 

_By default, Filament will send flash notifications via the Laravel session. However, you may wish that your notifications are "broadcast" to a user in real-time, instead. This could be used to send a temporary success notification from a queued job after it has finished processing._

What are "flash notification", exactly? Session based means they can not be broadcast to other users?
What does "broadcast to a user" mean, exactly? Doesn't broadcast imply users are multiple?
What does rea-time mean, exactly? Isn't a queued notification delayed rather than real-time?

I had the docs checked by a second person and he also found himself confused about the same points.

Hope this helps!

**Reply by pxlrbt** [reply]
Regarding the naming:
- Flash is referring to the Laravel `session()->flash()` concept: https://laravel.com/docs/master/session#flash-data
- Broadcast is referring to Laravel Broadcasting concept: https://laravel.com/docs/master/broadcasting#main-content

**Reply by pxlrbt** [reply]
Maybe we can combine the "Receiving notifications > Echo" part and the "Broadcasting" part?

**Reply by danharrin** [reply]
Echo also relates to database notifications though, which are not related to broadcasting

**Reply by KONEY** [reply]
> Regarding the naming:
> 
> * Flash is referring to the Laravel `session()->flash()` concept: https://laravel.com/docs/master/session#flash-data
> * Broadcast is referring to Laravel Broadcasting concept: https://laravel.com/docs/master/broadcasting#main-content

This is what I eventually figured out later, after configuring everything. But reading the two related sentences from the docs suggests that "Filament will send flash notification and these can be broadcast". Which it's not the case because broadcast notification will not be delivered to the user in the same exact way.

**Comment by Zeeshan-H** [comment]
Hi @danharrin  i am working on a project in filament and want to know that how can i show form fields on the basis of the current page?

**Reply by danharrin** [reply]
https://filamentphp.com/docs/2.x/admin/resources/getting-started#hiding-components-contextually

**Reply by Zeeshan-H** [reply]
Thanks @danharrin

**Comment by nicolaibaaring** [comment]
I just went through a nightmare setting up file uploads to AWS S3. It certainly needs more explanation in the documentation.
There is no explanation of the need to use the methods presented in the docs - only that they exist.
```php
FileUpload::make('attachment')
    ->disk('s3')
    ->directory('form-attachments')
    ->visibility('private')
```

The process should be explained in the docs.
1. Filament doesn't respect the Laravel filesystem driver setting. And since it doesn't seem logical to have a default of setting a driver at runtime, it has to be set in `config/filament.php` which requires publishing of the file. Or alternatively (and better) `FILAMENT_FILESYSTEM_DRIVER="${FILESYSTEM_DISK}"` has to be added to  `.env`, which is again not described. 
2. The actual upload process is a combination of Livewire upload which moves the file to `livewire-tmp` and Filament's additional process, which moves the file to the assigned directory before the action method is called.
3. The visibility setting is required on a non-public bucket for the file to actually be moved to the assigned directory. Otherwise the file is only deleted from `livewire-tmp` and not moved to the directory.

This suggests to improve the docs to include explanations of what is actually needed to make uploads work.

**Reply by nicolaibaaring** [reply]
Furthermore, is should be noted in the docs (as a warning) that it's not possible to add an Upload field to a modal, custom page or anything else than a standard page, which is very unfortunate for import scripts, as they do not fit onto Create or Edit pages.

https://github.com/filamentphp/filament/issues/5737

**Reply by danharrin** [reply]
That isn't true, uploads should work in any form, including in modals and custom pages

**Reply by danharrin** [reply]
The issue you linked is just a console error which affects modals, not custom pages, AFAIK it doesnt even prevent the file from being uploaded, its just a nuisance for logging

**Reply by nicolaibaaring** [reply]
OK. I'll check again and report back. Thanks.

**Reply by nicolaibaaring** [reply]
Yes, you are right. The upload does work in modals. It throws the error but the upload actually completes.

**Comment by StaffCollab** [comment]
Hi Everyone.

I am concerned with a filament concept that I need clarification on.

I've been using filament for a long time. Everything is fantastic, but the actions are all over the place.

Because all Actions seem equal (infolist, form, table, form component, global search... and there is probably even more), they all have similar syntax but not identical capabilities. This has tripped me up quite a few times.

I will give an example: infolists cannot open action modals if they live in a modal themselves, but form actions can.
Another example is that some actions require you to use `using()` to overwrite the default action; others don't have `using()` but have `action()` methods. Why shouldn't we always be able to use the `action()` to write or overwrite an action?
I have some more examples that almost made me cry... because it was only after I configured the whole view page as a slideover to give a nice effect that I noticed that my action forms wouldn't work on it.

In essence, I would love to create dedicated action classes (by the way, there are no docs on how to do it) so I can reuse them in resource pages, tables, and relation managers. However, since actions of every type perform differently, it is almost impossible.

I would prefer always to use the same action types, even if that means sometimes giving up on some magic dependency auto-injection.

I might be wrong or opinionated, and it might be written in the wrong thread. Please forgive me.
I love filament. I would still live it more.

Regards to everyone
SC

**Reply by danharrin** [reply]
Its all done in v4. Except the using() stuff - action() works to override any action function completely, but using() is used to override the tiny part of the action function that 99% of people actually override, its just a DX thing but action() still works

**Reply by StaffCollab** [reply]
Thanks, @danharrin for your wisdom and this awesome stuff you're giving us free of charge.

Is there a time plan (even if it's not a commitment) for the v4 release? Will this work in conjunction with Livewires versioning?

**Reply by danharrin** [reply]
It will probably not follow LW versioning, I don't think they are planning v4 this year and we would like to release this year.

**Comment by MACscr** [comment]
I recommend listening to the recent podcast.- MarkOn Apr 11, 2024, at 9:05 PM, StaffCollab ***@***.***> wrote:﻿
Hi Everyone.
I am concerned with a filament concept that I need clarification on.
I've been using filament for a long time. Everything is fantastic, but the actions are all over the place.
Because all Actions seem equal (infolist, form, table, form component, global search... and there is probably even more), they all have similar syntax but not identical capabilities. This has tripped me up quite a few times.
I will give an example: infolists cannot open action modals if they live in a modal themselves, but form actions can.
Another example is that some actions require you to use using() to overwrite the default action; others don't have using() but have action() methods. Why shouldn't we always be able to use the action() to write or overwrite an action?
I have some more examples that almost made me cry... because it was only after I configured the whole view page as a slideover to give a nice effect that I noticed that my action forms wouldn't work on it.
In essence, I would love to create dedicated action classes (by the way, there are no docs on how to do it) so I can reuse them in resource pages, tables, and relation managers. However, since actions of every type perform differently, it is almost impossible.
I would prefer always to use the same action types, even if that means sometimes giving up on some magic dependency auto-injection.
I might be wrong or opinionated, and it might be written in the wrong thread. Please forgive me.
I love filament. I would still live it more.
Regards to everyone
SC
